core/src/HIP/Kokkos_HIP_Locks.hpp

```diff
@@ -0,0 +1,180 @@
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_HIP_LOCKS_HPP
+#define KOKKOS_HIP_LOCKS_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#ifdef KOKKOS_ENABLE_HIP
+
+#include <cstdint>
+
+#include <HIP/Kokkos_HIP_Error.hpp>
+
+// FIXME_HIP We cannot use global variables defined in a namespace
+
+struct HIPLockArrays {
+  std::int32_t* atomic;
+  std::int32_t* scratch;
+  std::int32_t n;
+};
+
+/// \brief This global variable in Host space is the central definition
+///        of these arrays.
+extern HIPLockArrays g_host_hip_lock_arrays;
+
+namespace Kokkos {
+namespace Impl {
+
+/// \brief After this call, the g_host_hip_lock_arrays variable has
+///        valid, initialized arrays.
+///
+/// This call is idempotent.
+void initialize_host_hip_lock_arrays();
+
+/// \brief After this call, the g_host_hip_lock_arrays variable has
+///        all null pointers, and all array memory has been freed.
+///
+/// This call is idempotent.
+void finalize_host_hip_lock_arrays();
+
+}  // namespace Impl
+}  // namespace Kokkos
+
+#if defined(__HIPCC__)
+
+/// \brief This global variable in HIP space is what kernels use
+///        to get access to the lock arrays.
+///
+/// When relocatable device code is enabled, there can be one single
+/// instance of this global variable for the entire executable,
+/// whose definition will be in Kokkos_HIP_Locks.cpp (and whose declaration
+/// here must then be extern.
+/// This one instance will be initialized by initialize_host_HIP_lock_arrays
+/// and need not be modified afterwards.
+///
+/// When relocatable device code is disabled, an instance of this variable
+/// will be created in every translation unit that sees this header file.
+/// Since the Kokkos_HIP_Locks.cpp translation unit cannot initialize the
+/// instances in other translation units, we must update this HIP global
+/// variable based on the Host global variable prior to running any kernels
+/// that will use it.
+/// That is the purpose of the KOKKOS_ENSURE_HIP_LOCK_ARRAYS_ON_DEVICE macro.
+__device__
+// TODO define macro
+#ifdef KOKKOS_ENABLE_HIP_RELOCATABLE_DEVICE_CODE
+    __constant__ extern
+#endif
+    HIPLockArrays g_device_hip_lock_arrays;
+
+#define HIP_SPACE_ATOMIC_MASK 0x1FFFF
+
+namespace Kokkos {
+namespace Impl {
+
+/// \brief Acquire a lock for the address
+///
+/// This function tries to acquire the lock for the hash value derived
+/// from the provided ptr. If the lock is successfully acquired the
+/// function returns true. Otherwise it returns false.
+__device__ inline bool lock_address_hip_space(void* ptr) {
```

Comment:
AFAICT, this function is not used in this pull request at all. Apart from the definition here, I only find one more occurrence in `Kokkos_HIP_Atomic.hpp` in commented code.
Am I missing something here or should this be in a separate pull request?

Reply:
I still think that it should be in a separate pull request then.